scene/3d/spatial.cpp

```diff
@@ -732,6 +735,36 @@ Ref<World> Spatial::get_world() const {
 	return data.viewport->find_world();
 }
 
+void Spatial::_update_visible_in_tree() {
+	Spatial *parent = get_parent_spatial();
+
+	bool propagate_visible = parent ? parent->data.visible_in_tree : true;
+
+	// Only propagate visible when entering tree if we are visible.
+	propagate_visible &= is_visible();
+
+	_propagate_visible_in_tree(propagate_visible);
+}
+
+void Spatial::_propagate_visible_in_tree(bool p_visible_in_tree) {
+	// If any node is invisible, the propagation changes to invisible below.
+	p_visible_in_tree &= is_visible();
+
+	// No change.
+	if (data.visible_in_tree == p_visible_in_tree) {
+		return;
+	}
+
+	data.visible_in_tree = p_visible_in_tree;
+
+	for (int32_t n = 0; n < get_child_count(); n++) {
+		Spatial *s = Object::cast_to<Spatial>(get_child(n));
+		if (s) {
+			s->_propagate_visible_in_tree(p_visible_in_tree);
+		}
+	}
+}
```

Comment:
Hrm, perhaps this should just be merged into the children iteration we're already doing for `_propagate_visible_in_tree`?

Reply:
We discussed this in RocketChat; merging these is something we want to pursue, but `_propagate_visibility_changed` is currently using linked list iteration. Lawnjelly look into profiling this before making changes.